src-input/duk_bi_encoding.c

```diff
@@ -0,0 +1,348 @@
+			codepoint = 0x010000L | ((enc_ctx->lead - 0xd800L) << 10) | (codepoint - 0xdc00L);
+			enc_ctx->out += duk_unicode_encode_xutf8(codepoint, enc_ctx->out);
+			enc_ctx->lead = 0x0000L;
+		} else {
+			/* unpaired low surrogate */
+			enc_ctx->out += duk_unicode_encode_xutf8(DUK_UNICODE_CP_REPLACEMENT_CHARACTER, enc_ctx->out);
+		}
+	} else {
+		if (enc_ctx->lead != 0x0000L) {
+			/* unpaired high surrogate */
+			enc_ctx->out += duk_unicode_encode_xutf8(DUK_UNICODE_CP_REPLACEMENT_CHARACTER, enc_ctx->out);
+			enc_ctx->lead = 0x0000L;
+		}
+		enc_ctx->out += duk_unicode_encode_xutf8(codepoint, enc_ctx->out);
+	}
+}
+
+DUK_INTERNAL duk_ret_t duk_bi_textencoder_constructor(duk_context *ctx) {
+	/*
+	 *  TextEncoder currently requires no persistent state, so this constructor
+	 *  does nothing on purpose.
+	 */
+
+	if (!duk_is_constructor_call(ctx)) {
+        duk_error(ctx, DUK_ERR_TYPE_ERROR, "constructor must use 'new'");
+    }
+	return 0;
+}
+
+DUK_INTERNAL duk_ret_t duk_bi_textencoder_prototype_encoding_getter(duk_context *ctx) {
+	duk_push_string(ctx, "utf-8");
+	return 1;
+}
+
+DUK_INTERNAL duk_ret_t duk_bi_textencoder_prototype_encode(duk_context *ctx) {
+	duk__encode_context enc_ctx;
+	duk_size_t len;
+	duk_uint8_t *output;
+
+	if (duk_get_top(ctx) >= 1) {
+		duk_dup(ctx, 0);
+		duk_to_string(ctx, -1);
+	} else {
+		duk_push_string(ctx, "");
+	}
+
+	len = duk_get_length(ctx, -1);
+	if (len >= DUK_HBUFFER_MAX_BYTELEN / 3) {
+		duk_error(ctx, DUK_ERR_TYPE_ERROR, "input string too long");
+	}
+
+	/* XXX: The buffer allocation strategy here (3x input length) is quite
+	 * inefficient.  Maybe switch to a chunk-based strategy?
+	 */
+
+	output = (duk_uint8_t *) duk_push_dynamic_buffer(ctx, len * 3);
+	enc_ctx.lead = 0x0000L;
+	enc_ctx.out = output;
+	duk_decode_string(ctx, -2, duk__utf8_encode_char, (void *) &enc_ctx);
+	if (enc_ctx.lead != 0x0000L) {
+		/* unpaired high surrogate at end of string */
+		enc_ctx.out += duk_unicode_encode_xutf8(DUK_UNICODE_CP_REPLACEMENT_CHARACTER, enc_ctx.out);
+	}
+#if defined(DUK_USE_BUFFEROBJECT_SUPPORT)
+	duk_push_buffer_object(ctx, -1, 0, (duk_size_t) (enc_ctx.out - output), DUK_BUFOBJ_UINT8ARRAY);
+#else
+	/* shrink buffer to include only the encoded text */
+	duk_resize_buffer(ctx, -1, (duk_size_t) (enc_ctx.out - output));
+#endif
+	return 1;
+}
+
+DUK_INTERNAL duk_ret_t duk_bi_textdecoder_constructor(duk_context *ctx) {
+	duk__decode_context *dec_ctx;
+	duk_bool_t fatal = 0;
+	duk_bool_t ignore_bom = 0;
+	duk_idx_t nargs;
+
+	nargs = duk_get_top(ctx);
+	if (!duk_is_constructor_call(ctx)) {
+        duk_error(ctx, DUK_ERR_TYPE_ERROR, "constructor must use 'new'");
+    }
+	if (nargs >= 1) {
+		/* XXX: for now, label is ignored */
+		duk_to_string(ctx, 0);
+	}
+	if (nargs >= 2 && !duk_is_null_or_undefined(ctx, 1)) {
+		if (duk_get_prop_string(ctx, 1, "fatal")) {
+			fatal = duk_to_boolean(ctx, -1);
+		}
+		if (duk_get_prop_string(ctx, 1, "ignoreBOM")) {
+			ignore_bom = duk_to_boolean(ctx, -1);
+		}
```

Comment:
In any case, my point was simply that if there are ~5 call sites, an internal helper will reduce code footprint and also make the call sites shorter (though not necessarily more easy to read unless the helper is named well). For internal helpers long term viability or consistency are not critical because they can be reworked at will.